### PR TITLE
Group transitions

### PR DESCRIPTION
**Pull request by zenovich**

### CODE REVIEW

File: app/database/group_group_store_transitions.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+			var maxChildOrder struct{ MaxChildOrder int64 }
+			mustNotBeError(s.WithWriteLock().
+				Select("IFNULL(MAX(iChildOrder), 0)").
+				Where("idGroupParent = ?", parentGroupID).Scan(&maxChildOrder).Error())
+
+			insertQuery := "INSERT INTO groups_groups (ID, idGroupParent, idGroupChild, sType, iChildOrder, sStatusDate) VALUES " +
+				strings.Repeat("(?, ?, ?, ?, ?, NOW()), ", len(idsToInsert)-1) +
+				"(?, ?, ?, ?, ?, NOW())"
+			mustNotBeError(s.retryOnDuplicatePrimaryKeyError(func(db *DB) error {
+				values := make([]interface{}, 0, len(idsToInsert)*4)
```
Comment (smadbe) [inline app/database/group_group_store_transitions.go]: don't you add 6 values for each?

Comment (zenovich) [reply app/database/group_group_store_transitions.go]: No, just 5 :) Fixed this.

File: app/database/group_group_store_transitions.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+type GroupGroupTransitionAction int
+
+const (
+	AdminCreatesInvitation GroupGroupTransitionAction = iota
+	UserCreatesRequest
+	UserAcceptsInvitation
+	AdminAcceptsRequest
+	UserRefusesInvitation
+	AdminRefusesRequest
+	AdminRemovesChild
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/group_group_store_transitions.go]: added comments

File: app/database/group_group_store_transitions.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	AdminCancelsInvitation
+	UserLeavesGroup
+	UserCancelsRequest
+	AdminAddsDirectRelation
+)
+
+type groupGroupTransitionRule struct {
+	UpdateFromType map[GroupGroupType]bool
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/group_group_store_transitions.go]: added comments

File: app/database/group_group_store_transitions.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+			RequestRefused:     Direct,
+			Removed:            Direct,
+			Left:               Direct,
+			Direct:             Direct,
+		},
+	},
+}
+
+type GroupGroupTransitionResults struct {
+	Success   map[int64]bool `json:"success,omitempty"`
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/group_group_store_transitions.go]: removed 'json' tags

**Pull request merged**